odbc/src/api/data.rs

```diff
@@ -196,9 +198,58 @@ impl IntoCData for Bson {
     fn to_datetime(&self) -> Result<DateTime<Utc>> {
         match self {
             Bson::DateTime(d) => Ok((*d).into()),
-            Bson::String(s) => Utc
-                .datetime_from_str(s, "%FT%H:%M:%S%.3fZ")
-                .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone())),
+            Bson::String(s) => {
+                if s.contains('T') {
+                    // received a datetime string
+                    Utc.datetime_from_str(s, "%FT%H:%M:%S%.3fZ")
+                        .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone()))
+                } else if s.contains('-') {
+                    // received a date string
+                    let date = NaiveDate::parse_from_str(s, "%F")
+                        .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone()))?;
+                    Ok(DateTime::<Utc>::from_utc(date.and_hms(0, 0, 0), Utc))
+                } else {
+                    // received a time string -- should return utc today + time
+                    let time = NaiveTime::parse_from_str(s, "%H:%M:%S%.3f")
+                        .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone()))?;
+                    Ok(DateTime::<Utc>::from_utc(
+                        NaiveDateTime::new(Utc::today().naive_utc(), time),
+                        Utc,
+                    ))
+                }
+            }
+            o => Err(ODBCError::RestrictedDataType(o.to_type_str(), DATETIME)),
+        }
+    }
+
+    fn to_date(&self) -> Result<NaiveDate> {
+        match self {
+            Bson::DateTime(d) => Ok((*d).to_chrono().date_naive()),
+            Bson::String(s) => {
+                let fmt = if s.contains('T') {
+                    "%FT%H:%M:%S%.3fZ"
+                } else {
+                    "%F"
+                };
+                NaiveDate::parse_from_str(s, fmt)
+                    .map_err(|_| ODBCError::InvalidDatetimeFormat(s.clone()))
```

Comment:
I might be missing it, but are we handling cases where the time portion is non-zero?

From the spec
> Data value is a valid timestamp-value; time portion is zero[a] -> Data

> Data value is a valid timestamp-value; time portion is nonzero[a],[c] -> Truncated Data

Reply:
you're right (here and below)! will address